File: src/main/java/com/wikia/tibia/objects/Book.java
package com.wikia.tibia.objects;

import com.wikia.tibia.enums.BookType;
import com.wikia.tibia.enums.Status;
import com.wikia.tibia.utils.ObjectUtils;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class Book extends WikiObject {

    private static final Logger LOG = LoggerFactory.getLogger(Book.class);

    private BookType booktype;
    private BookType booktype2;
    private BookType booktype3;
    private BookType booktype4;
    private BookType booktype5;
    private BookType booktype6;
    private BookType booktype7;
    private BookType booktype8;
    private String title;
    private String pagename;
    private String location;
    private String location2;
    private String location3;
    private String location4;
    private String location5;
    private String location6;
    private String location7;
    private String location8;
    private String blurb;
    private String author;
    private String returnpage;
    private String returnpage2;
    private String returnpage3;
    private String returnpage4;
    private String returnpage5;
    private String returnpage6;
    private String returnpage7;
    private String returnpage8;
    private String returnpage9;
    private String returnpage10;
    private String returnpage11;
    private String returnpage12;
    private String returnpage13;
    private String returnpage14;
    private String returnpage15;
    private String returnpage16;
    private String prevbook;
    private String nextbook;
    private String relatedpages;
    private String text;
    private String text2;
    private String text3;
    private String text4;
    private String text5;
    private String text6;
    private String text7;
    private String text8;
    private String implemented2;
    private String implemented3;
    private String implemented4;
    private String implemented5;
    private String implemented6;
    private String implemented7;
    private String implemented8;

    @Builder
    public Book(String name, String implemented, String notes, String history, Status status, BookType booktype, String title,
                String pagename, String location, String blurb, String author, String returnpage, String prevbook,
                String nextbook, String relatedpages, String text) {
        super(name, null, null, null, implemented, notes, history, status);
        this.booktype = booktype;
        this.booktype2 = null;
        this.booktype3 = null;
        this.booktype4 = null;
        this.booktype5 = null;
        this.booktype6 = null;
        this.booktype7 = null;
        this.booktype8 = null;
        this.title = title;
        this.pagename = pagename;
        this.location = location;
        this.location2 = null;
        this.location3 = null;
        this.location4 = null;
        this.location5 = null;
        this.location6 = null;
        this.location7 = null;
        this.location8 = null;
        this.blurb = blurb;
        this.author = author;
        this.returnpage = returnpage;
        this.returnpage2 = null;
        this.returnpage3 = null;
        this.returnpage4 = null;
        this.returnpage5 = null;
        this.returnpage6 = null;
        this.returnpage7 = null;
        this.returnpage8 = null;
        this.returnpage9 = null;
        this.returnpage10 = null;
        this.returnpage11 = null;
        this.returnpage12 = null;
        this.returnpage13 = null;
        this.returnpage14 = null;
        this.returnpage15 = null;
        this.returnpage16 = null;
        this.prevbook = prevbook;
        this.nextbook = nextbook;
        this.relatedpages = relatedpages;
        this.text = text;
        this.text2 = null;
        this.text3 = null;
        this.text4 = null;
        this.text5 = null;
        this.text6 = null;
        this.text7 = null;
        this.text8 = null;
        this.implemented2 = null;
        this.implemented3 = null;
        this.implemented4 = null;
        this.implemented5 = null;
        this.implemented6 = null;
        this.implemented7 = null;
        this.implemented8 = null;
    }

    @Override
    public String getName() {
        return getPagename();
    }

    @Override
    public void setDefaultValues() {

        if (ObjectUtils.isEmpty(getImplemented())) {
            setImplemented("?");
        }

        if (ObjectUtils.isEmpty(booktype)) {
            LOG.warn("Book '{}' has no booktype set", getName());
        }

        if (ObjectUtils.isEmpty(title)) {
            title = "?";
            LOG.warn("Book '{}' has no title set", getName());
        }

        if (ObjectUtils.isEmpty(pagename)) {
            pagename = "?";
            LOG.warn("Book '{}' has no pagename set", getName());
        }

        if (ObjectUtils.isEmpty(location)) {
            location = "?";
        }

        if (ObjectUtils.isEmpty(blurb)) {
            blurb = "?";
        }

        if (ObjectUtils.isEmpty(author)) {
            author = "?";
        }

        if (ObjectUtils.isEmpty(text)) {
            LOG.warn("Book '{}' has no text set", getName());
        }
    }
}

File: src/main/kotlin/com/wikia/tibia/objects/Book.kt
package com.wikia.tibia.objects

import com.wikia.tibia.enums.BookType
import com.wikia.tibia.enums.Status
import com.wikia.tibia.utils.ObjectUtils.isEmpty
import org.slf4j.LoggerFactory

data class Book(
  val name: String,
  val implemented: String?,
  val notes: String,
  val history: String,
  val status: Status,
  val booktype: BookType,
  val booktype2: BookType,
  val booktype3: BookType,
  val booktype4: BookType,
  val booktype5: BookType,
  val booktype6: BookType,
  val booktype7: BookType,
  val booktype8: BookType,
  val title: String?,
  val pagename: String?,
  val location: String?,
  val location2: String,
  val location3: String,
  val location4: String,
  val location5: String,
  val location6: String,
  val location7: String,
  val location8: String,
  val blurb: String?,
  val author: String?,
  val returnpage: String,
  val returnpage2: String,
  val returnpage3: String,
  val returnpage4: String,
  val returnpage5: String,
  val returnpage6: String,
  val returnpage7: String,
  val returnpage8: String,
  val returnpage9: String,
  val returnpage10: String,
  val returnpage11: String,
  val returnpage12: String,
  val returnpage13: String,
  val returnpage14: String,
  val returnpage15: String,
  val returnpage16: String,
  val prevbook: String,
  val nextbook: String,
  val relatedpages: String,
  val text: String,
  val text2: String,
  val text3: String,
  val text4: String,
  val text5: String,
  val text6: String,
  val text7: String,
  val text8: String,
  val implemented2: String,
  val implemented3: String,
  val implemented4: String,
  val implemented5: String,
  val implemented6: String,
  val implemented7: String,
  val implemented8: String
) : WikiObject() {

  override fun setDefaultValues() {
    if (isEmpty(implemented)) {
      this.copy(implemented = "?")
    }
    if (isEmpty(booktype)) {
      logger.warn("Book '{}' has no booktype set", name)
    }
    if (isEmpty(title)) {
      this.copy(title = "?")
      logger.warn("Book '{}' has no title set", name)
    }
    if (isEmpty(pagename)) {
      this.copy(pagename = "?")
      logger.warn("Book '{}' has no pagename set", name)
    }
    if (isEmpty(location)) {
      this.copy(location = "?")
    }
    if (isEmpty(blurb)) {
      this.copy(blurb = "?")
    }
    if (isEmpty(author)) {
      this.copy(author = "?")
    }
    if (isEmpty(text)) {
      logger.warn("Book '{}' has no text set", name)
    }
  }

  companion object {
    private val logger = LoggerFactory.getLogger(Book::class.java)
  }
}
